src/filter_tree/filter_tree.c

```diff
@@ -7,19 +7,58 @@
 #include "../value.h"
 #include "filter_tree.h"
 #include "../util/arr.h"
+#include "../ast/ast_shared.h"
 #include <assert.h>
 
+/* forward declarations */
+void _FilterTree_DeMorgan(FT_FilterNode **root, uint negate_count);
+
 static inline FT_FilterNode *LeftChild(const FT_FilterNode *node) {
 	return node->cond.left;
 }
 static inline FT_FilterNode *RightChild(const FT_FilterNode *node) {
 	return node->cond.right;
 }
 
+/* Returns the negated operator of given op.
+ * for example NOT(a > b) === a <= b */
+static AST_Operator _NegateOperator(AST_Operator op) {
+	switch(op) {
+	case OP_AND:
+		return OP_OR;
+	case OP_OR:
+		return OP_AND;
+	case OP_EQUAL:
+		return OP_NEQUAL;
+	case OP_NEQUAL:
+		return OP_EQUAL;
+	case OP_LT:
+		return OP_GE;
+	case OP_GT:
+		return OP_LE;
+	case OP_LE:
+		return OP_GT;
+	case OP_GE:
+		return OP_LT;
+	default:
+		assert(false);
+	}
+}
+
+static void _NegateExpression(AR_ExpNode **exp) {
+	AR_ExpNode *root = AR_EXP_NewOpNode("not", 1);
+	root->op.children[0] = *exp;
+	*exp = root;
+}
+
 int IsNodePredicate(const FT_FilterNode *node) {
 	return node->t == FT_N_PRED;
 }
 
+int IsNodeExpression(const FT_FilterNode *node) {
```

Comment:
Unused, but a sensible interface to have - your call on keeping it or not.

Reply:
Removed